src/lib.rs

```diff
@@ -19,78 +19,37 @@ fn _byte_pair_merge(
     ranks: &HashMap<Vec<u8>, Rank>,
     piece: &[u8],
 ) -> Vec<(usize, Rank)> {
-    // This is a vector of (start, rank).
-    // The rank is of the byte pair starting at position start.
-    // The rank of the last item in the vector is not a valid value.
-    let mut parts: Vec<(usize, Rank)> = (0..piece.len() + 1).map(|i| (i, Rank::MAX)).collect();
-
-    let get_rank = {
-        #[inline(always)]
-        |parts: &Vec<(usize, Rank)>, start_idx: usize, skip: usize| {
-            if (start_idx + skip + 2) < parts.len() {
-                ranks
-                    .get(&piece[parts[start_idx].0..parts[start_idx + skip + 2].0])
-                    .copied()
-            } else {
-                None
-            }
-        }
+    let get_rank = |parts: &Vec<(usize, _)>, start_idx: usize, end_idx: usize| {
+        *parts.get(end_idx)
+            .map(|e| parts.get(start_idx).unwrap().0..e.0)
+            .and_then(|r| piece.get(r))
+            .filter(|p| p.len() < piece.len())
+            .and_then(|p| ranks.get(p))
+            .unwrap_or(&Rank::MAX)
     };
 
-    // We look up the ranks once in the beginning and iteratively update
-    // them during each merge, which reduces the number of rank lookups.
-    for i in 0..parts.len() - 2 {
-        match get_rank(&parts, i, 0) {
-            Some(rank) => {
-                // Rank::MAX is a sentinel value and cannot be a valid rank
-                debug_assert!(rank != Rank::MAX);
-                parts[i].1 = rank;
-            }
-            None => {
-                continue;
-            }
-        };
-    }
-
-    // If you have n parts and m merges, this does O(mn) work.
-    // We could do something with a heap and do O(m log n) work.
-    // It is important to consider that n is often small (<100), and as such
-    // the cache-locality benefits outweigh the algorithmic complexity downsides
-    // of the `parts` vector data structure above.
-
-    // Note that we hash bytes, not token pairs. As long as we train BPE the way we
-    // currently do, this is equivalent. An easy way to break this would be to decouple
-    // merge priority from token index or to prevent specific token merges.
-    loop {
-        if parts.len() == 1 {
-            break;
+    let (mut min_rank_index, mut min_rank) = (0, Rank::MAX);
+    let mut parts = Vec::with_capacity(piece.len() + 1);
+    for i in 0..piece.len() + 1 {
+        let part = (i, *piece.get(i..i + 2).and_then(|p| ranks.get(p)).unwrap_or(&Rank::MAX));
+        if part.1 < min_rank {
+            (min_rank_index, min_rank) = part;
```

Comment:
since we're already iterating, we might as well calculate the minimum - if the word consists of single characters (e.g. as foreign words often do) this enables us to avoid the loop completely